File: typescript/packages/core/src/workspace/snapshot/drift.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { OpRecord } from '../../observe/record.ts'
import type { FileStat } from '../../types.ts'
import { DriftPolicy } from '../../types.ts'
import type { MountEntry } from '../mount/mount.ts'

/**
 * Raised at load time when a remote resource's live fingerprint differs
 * from what was recorded in the snapshot.
 *
 * Indicates the underlying source has been modified since the snapshot
 * was taken, so reading current bytes would silently diverge from what
 * the original agent saw. Surface to the caller rather than mask.
 */
export class ContentDriftError extends Error {
  readonly path: string
  readonly snapshotFingerprint: string
  readonly liveFingerprint: string | null

  constructor(path: string, snapshotFingerprint: string, liveFingerprint: string | null) {
    const liveRepr = liveFingerprint === null ? '<missing>' : JSON.stringify(liveFingerprint)
    super(
      `${path}: snapshot fingerprint ${JSON.stringify(snapshotFingerprint)}, live ${liveRepr}; ` +
        'data on the underlying source has changed since the snapshot was taken',
    )
    this.name = 'ContentDriftError'
    this.path = path
    this.snapshotFingerprint = snapshotFingerprint
    this.liveFingerprint = liveFingerprint
  }
}

export interface FingerprintEntry {
  path: string
  mount_prefix: string
  fingerprint?: string | null
  revision?: string | null
}

interface RegistryLike {
  mountFor(path: string): MountEntry | null
  allMounts(): readonly MountEntry[]
}

/**
 * Fingerprint checks a load queued, drained on the first async op.
 *
 * `Workspace.load` records one entry per read whose snapshot manifest
 * carried a fingerprint but no stable revision (a pinned read needs no
 * check: the pin guarantees the bytes). The first `dispatch` or
 * `execute` drains them, so downstream code can rely on consistent
 * state. Mirrors the Python `DriftQueue` in `snapshot/drift.py`.
 */
export class DriftQueue {
  private entries: { path: string; fingerprint: string }[] = []
  private isPending = false

  get pending(): boolean {
    return this.isPending
  }

  /** Paths still queued for a check (audit surface). */
  get paths(): string[] {
    return this.entries.map((e) => e.path)
  }

  /** Drop any queued state (a re-install starts fresh). */
  clear(): void {
    this.entries = []
    this.isPending = false
  }

  queue(path: string, fingerprint: string): void {
    this.entries.push({ path, fingerprint })
    this.isPending = true
  }

  /**
   * Stat every queued path in parallel; throw on the first drift.
   * Subsequent calls are no-ops. Stats run concurrently so first-op
   * latency does not scale linearly with the number of recorded reads.
   */
  async drain(registry: RegistryLike, statFn: (path: string) => Promise<unknown>): Promise<void> {
    this.isPending = false
    if (this.entries.length === 0) return
    const pending = this.entries
    this.entries = []
    const results = await Promise.allSettled(
      pending.map((p) => checkDrift(registry, statFn, p.path, p.fingerprint)),
    )
    for (const r of results) {
      if (r.status === 'rejected') throw r.reason as Error
    }
  }
}

/**
 * Walk a loaded snapshot's fingerprint manifest. For entries with a
 * revision, install the pin on the owning mount so replay reads pin to
 * that revision. For fingerprint-only entries, queue the path on the
 * drift queue. OFF skips the checks and evicts the snapshot cache for
 * fingerprinted paths so reads serve current state.
 *
 * Idempotent: clears queued state before installing. Called from
 * `Workspace.fromState`.
 */
export function installDriftState(
  registry: RegistryLike,
  cache: { remove(key: string): Promise<void> },
  drift: DriftQueue,
  state: { fingerprints?: FingerprintEntry[]; live_only_mounts?: string[] },
  policy: DriftPolicy,
): void {
  drift.clear()
  const entries = state.fingerprints ?? []
  if (entries.length === 0) return
  if (policy === DriftPolicy.OFF) {
    for (const e of entries) {
      void cache.remove(e.path)
    }
    return
  }
  for (const e of entries) {
    const mount = registry.mountFor(e.path)
    if (mount === null) continue
    if (e.revision !== undefined && e.revision !== null) {
      mount.revisions.set(e.path, e.revision)
      continue
    }
    if (e.fingerprint !== undefined && e.fingerprint !== null) {
      drift.queue(e.path, e.fingerprint)
    }
  }
  const liveOnly = state.live_only_mounts ?? []
  if (liveOnly.length > 0) {
    console.warn(
      `Workspace.load: ${String(liveOnly.length)} mount(s) opt out of snapshot replay; ` +
        `reads against them will serve current state with no drift detection: ` +
        liveOnly.join(', '),
    )
  }
}

/**
 * Walk recorded ops and emit one entry per distinct read on a
 * snapshot-capable mount.
 *
 * Pure aggregation over `records`. Each read carries the `fingerprint`
 * and/or `revision` the backend returned at the moment the agent read
 * the bytes (populated from the GET response, not a fresh stat at
 * snapshot time). This avoids the race where the upstream changes
 * between read and snapshot.
 *
 * Skips paths whose owning mount has `supportsSnapshot=false` (live-only
 * backends like Gmail/Slack/Linear) and reads where the backend returned
 * neither marker.
 */
export function captureFingerprints(
  records: readonly OpRecord[],
  registry: RegistryLike,
): FingerprintEntry[] {
  const seen = new Set<string>()
  const out: FingerprintEntry[] = []
  for (const rec of records) {
    if (rec.op !== 'read' || seen.has(rec.path)) continue
    if (rec.fingerprint === null && rec.revision === null) continue
    seen.add(rec.path)
    const mount = registry.mountFor(rec.path)
    if (mount === null) continue
    if (mount.resource.supportsSnapshot !== true) continue
    const entry: FingerprintEntry = { path: rec.path, mount_prefix: mount.prefix }
    if (rec.fingerprint !== null) entry.fingerprint = rec.fingerprint
    if (rec.revision !== null) entry.revision = rec.revision
    out.push(entry)
  }
  return out
}

/**
 * Return mount prefixes whose resource opts out of snapshot replay.
 *
 * These mounts will serve current state at load time with no drift
 * detection. Surfaced in the snapshot manifest so the load layer can
 * log them and so users can audit which paths are non-replayable.
 */
export function liveOnlyMountPrefixes(registry: RegistryLike): string[] {
  const out: string[] = []
  for (const m of registry.allMounts()) {
    if (m.prefix === '/dev/' || m.prefix === '/.bash_history/' || m.prefix === '/bin/') continue
    if (m.resource.supportsSnapshot !== true) out.push(m.prefix)
  }
  return out
}

/**
 * Stat `path` and throw {@link ContentDriftError} if the live fingerprint
 * does not match `recorded`. No-op if the mount cannot be resolved or the
 * resource cannot fingerprint.
 *
 * The caller provides `statFn` (typically a thin wrapper over
 * {@link Workspace.dispatch}) so that drift.ts stays decoupled from the
 * workspace's op-resolution machinery.
 */
export async function checkDrift(
  registry: RegistryLike,
  statFn: (path: string) => Promise<unknown>,
  path: string,
  recorded: string,
): Promise<void> {
  const mount = registry.mountFor(path)
  if (mount === null) return
  if (mount.resource.supportsSnapshot !== true) return
  let stat: FileStat
  try {
    stat = (await statFn(path)) as FileStat
  } catch (err) {
    if ((err as { code?: string } | null)?.code === 'ENOENT') {
      throw new ContentDriftError(path, recorded, null)
    }
    throw err
  }
  const live = stat.fingerprint
  if (live === null) return
  if (live !== recorded) throw new ContentDriftError(path, recorded, live)
}

File: python/mirage/workspace/snapshot/drift.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import logging
from typing import Any

from mirage.types import DriftPolicy, FingerprintKey

logger = logging.getLogger(__name__)


class ContentDriftError(Exception):
    """Raised at load time when a remote resource's live fingerprint
    differs from what was recorded in the snapshot.

    Indicates the underlying source has been modified since the snapshot
    was taken, so reading current bytes would silently diverge from what
    the original agent saw. Surface to the caller rather than mask.

    Attributes:
        path (str): Virtual path that drifted.
        snapshot_fingerprint (str): Recorded marker.
        live_fingerprint (str | None): Marker observed at load time.
    """

    def __init__(self, path: str, snapshot_fingerprint: str,
                 live_fingerprint: str | None) -> None:
        self.path = path
        self.snapshot_fingerprint = snapshot_fingerprint
        self.live_fingerprint = live_fingerprint
        live_repr = repr(
            live_fingerprint) if live_fingerprint is not None else "<missing>"
        super().__init__(
            f"{path}: snapshot fingerprint {snapshot_fingerprint!r}, "
            f"live {live_repr}; data on the underlying source has changed "
            "since the snapshot was taken")


class DriftQueue:
    """Fingerprint checks a load queued, drained on the first async op.

    ``Workspace.load`` records one entry per read whose snapshot
    manifest carried a fingerprint but no stable revision (a pinned
    read needs no check: the pin guarantees the bytes). The first
    ``dispatch`` or ``execute`` drains them, so downstream code can
    rely on consistent state.
    """

    def __init__(self) -> None:
        self._entries: list[tuple[str, str]] = []
        self._pending = False

    @property
    def pending(self) -> bool:
        return self._pending

    @property
    def paths(self) -> list[str]:
        """Paths still queued for a check (audit surface)."""
        return [path for path, _ in self._entries]

    def queue(self, path: str, fingerprint: str) -> None:
        """Record one path to check against its live source.

        Args:
            path (str): virtual path recorded in the snapshot.
            fingerprint (str): marker the snapshot recorded for it.
        """
        self._entries.append((path, fingerprint))
        self._pending = True

    async def drain(self, ws) -> None:
        """Stat every queued path in parallel; raise on the first drift.

        Subsequent calls are no-ops. Stats are issued with
        ``asyncio.gather`` so first-op latency does not scale linearly
        with the number of recorded reads.

        Args:
            ws: Workspace whose registry resolves the paths.

        Raises:
            ContentDriftError: a live fingerprint differs from the
                recorded one.
        """
        self._pending = False
        if not self._entries:
            return
        checks = [
            check_drift(ws, path, fingerprint)
            for path, fingerprint in self._entries
        ]
        self._entries.clear()
        results = await asyncio.gather(*checks, return_exceptions=True)
        for result in results:
            if isinstance(result, BaseException):
                raise result


def capture_fingerprints(ws) -> list[dict[str, Any]]:
    """Walk session ops and emit one entry per distinct read on a
    ``SUPPORTS_SNAPSHOT`` mount.

    Pure aggregation over ``ws._ops.records``. Each read ``OpRecord``
    carries the ``fingerprint`` and/or ``revision`` the backend
    returned at the moment the agent read the bytes (populated from
    the GET response, not a fresh stat at snapshot time). This avoids
    the race where the upstream changes between read and snapshot.

    Skips paths whose owning mount has ``SUPPORTS_SNAPSHOT=False``
    (live-only backends like Gmail/Slack/Linear) and reads where the
    backend returned neither marker.

    Args:
        ws: Workspace whose ops log to walk.

    Returns:
        list[dict]: One entry per (recorded, fingerprinted) path, with
        ``PATH``, ``MOUNT_PREFIX`` and at least one of ``FINGERPRINT``
        or ``REVISION``. Both may be present on versioned backends that
        return ETag and VersionId on every GET.
    """
    seen: set[str] = set()
    out: list[dict[str, Any]] = []
    for rec in ws._ops.records:
        if rec.op != "read" or rec.path in seen:
            continue
        if rec.fingerprint is None and rec.revision is None:
            continue
        seen.add(rec.path)
        try:
            mount = ws._registry.mount_for(rec.path)
        except ValueError:
            continue
        if not getattr(mount.resource, "SUPPORTS_SNAPSHOT", False):
            continue
        entry: dict[str, Any] = {
            FingerprintKey.PATH: rec.path,
            FingerprintKey.MOUNT_PREFIX: mount.prefix,
        }
        if rec.fingerprint is not None:
            entry[FingerprintKey.FINGERPRINT] = rec.fingerprint
        if rec.revision is not None:
            entry[FingerprintKey.REVISION] = rec.revision
        out.append(entry)
    return out


def install_fingerprints(ws, fingerprint_entries: list[dict[str, Any]],
                         drift_policy: DriftPolicy) -> None:
    """Install snapshot fingerprints/revisions onto a reconstructed ws.

    Revisions pin replay reads to exact backend versions; bare
    fingerprints queue an eager drift check. OFF evicts the snapshot
    cache for fingerprinted paths so reads serve current state.

    Args:
        ws: the reconstructed workspace to install onto.
        fingerprint_entries: entries from a snapshot's FINGERPRINTS.
        drift_policy: STRICT queues drift checks; OFF skips and evicts.
    """
    if drift_policy == DriftPolicy.OFF:
        if fingerprint_entries:
            ws._cache.evict_paths(f[FingerprintKey.PATH]
                                  for f in fingerprint_entries)
        return
    for f in fingerprint_entries:
        path = f[FingerprintKey.PATH]
        try:
            mount = ws._registry.mount_for(path)
        except ValueError:
            continue
        revision = f.get(FingerprintKey.REVISION)
        if revision is not None:
            mount.revisions[path] = revision
            continue
        fingerprint = f.get(FingerprintKey.FINGERPRINT)
        if fingerprint is not None:
            ws._drift.queue(path, fingerprint)


def live_only_mount_prefixes(ws) -> list[str]:
    """Return mount prefixes whose resource opts out of snapshot replay.

    These mounts will serve current state at load time with no drift
    detection. Surfaced in the snapshot manifest so the load layer can
    log them and so users can audit which paths are non-replayable.
    """
    out: list[str] = []
    for m in ws._registry.mounts():
        if m.prefix in {"/dev/", "/.bash_history/", "/bin/"}:
            continue
        if not getattr(m.resource, "SUPPORTS_SNAPSHOT", False):
            out.append(m.prefix)
    return out


async def check_drift(ws, path: str, recorded: str) -> None:
    """Stat `path` against its mount and raise ContentDriftError if the
    live fingerprint does not match `recorded`.

    No-op if the mount cannot be resolved or the resource cannot
    fingerprint (raises only on a real, observable mismatch).

    Args:
        ws: Workspace whose registry to consult.
        path (str): Virtual path to check.
        recorded (str): Fingerprint recorded at snapshot time.

    Raises:
        ContentDriftError: live fingerprint differs from recorded.
    """
    try:
        mount = ws._registry.mount_for(path)
    except ValueError:
        return
    if not getattr(mount.resource, "SUPPORTS_SNAPSHOT", False):
        return
    try:
        stat = await mount.execute_op("stat", path)
    except FileNotFoundError as exc:
        raise ContentDriftError(path, recorded, None) from exc
    live = getattr(stat, "fingerprint", None)
    if live is None:
        return
    if live != recorded:
        raise ContentDriftError(path, recorded, live)
